### ISSUE TITLE
Evaluating models created using python bindings in C# 

### ISSUE THREAD

**Issue by Troelsj** [Microsoft/CNTK#1050]
Hi,

I have a problem with evaluating models, created using the python bindings, through the C# evaluation interface.

If I follow the examples on saving a model it would look like:
``python 
        persist.save_model(z, os.path.join(model_path, network_name + "_{}.dnn".format(epoch)))
``

If I then load that model into a C# program in order to classify an image, it returns the error:

> No Output nodes found: Cannot determine Input node dimensions due to lack of Output nodes.
(are 'outputNodeNames' and/or 'OutputNodes' properly defined in the configuration file?)

(Both with my own models and the models from the examples)
Which seems like a simple error, but I can't find any explanation on how this should be defined in python, it works fine if I load a brainscript model into the program.

Using the python evaluation function works with the python model, something like:

``python
model = load_model("model.bin")
input_arg = model.arguments
test_image  = np.array(Image.open("image.png"), dtype = np.float32).T
results = np.squeeze(model.eval({ input_arg:test_image }))
``

But that doesn't work with a brainscript model, even if I change it to fit the extra inputs.

So are the models really not created equally, why is the resnet20 model created in python not the same as the one created in brainscript? 
Is there an easy way of defining outputnodes using the python bindings, such that it works with a C# evaluation program?

I am running cntk v2.0 beta3, built from source for GPU usage.

**Comment by zhouwangzw** [comment]
@Troelsj Yes, the Python API has a different naming schema than BrainScript. And currently it is not easy for users to retrieve node names from the Python model. We are currently working on it, and the C# Eval API for V2 library is also under development. 
For now, the following workaound might work for you. I am assuming that you are able to use model.CreateNetwork() to load the Python model into your C# program. If yes, by adding tracelevel=1 when creating a network, i.e. 
  `model.CreateNetwork(string.Format("traceLevel=1\n modelPath=\"{0}\"", modelFilePath), deviceId: -1);`
you should be able to see the output of "validating network" where all nodes are printed out. From there you might find the name of output nodes you want. 
After that, you need to change CreateNetwork() by passing the parameter "outputNodeNames" with the list of output node names:
  `var desiredOutputLayers = new List<string>() { "out.z", "out.h1" };
   model.CreateNetwork(string.Format("modelPath=\"{0}\"", modelFilePath), deviceId: -1, outputNodeNames: desiredOutputLayers); `  
Then you can call GetNodeDimensions to retrieve dimensions of input and output nodes, and then proceed with evaluation. 

Sorry for the inconvenience, and please let me know if you have any questions.
Thanks

**Comment by Troelsj** [comment]
@zhouwangzw Sorry for the late reply, but that indeed makes it possible to create the network, unfortunately I am still not able to actually evaluate with a model created in python.

The output lines from brainscript looks like:

>Validating --> out = Plus (out.PlusArgs[0], model.arrayOfFunctions[8].b) : [2 x *], [2] -> [2 x *]
>Validating --> ce = CrossEntropyWithSoftmax (labels, out) : [2 x *], [2 x *] -> [1]
>Validating --> errs = ClassificationError (labels, out) : [2 x *], [2 x *] -> [1]
>Validating --> inputs.inputs[2] = LearnableParameter() :  -> [1 x 1]
>Validating --> top5Errs = ClassificationError (labels, out, inputs.inputs[2]) : [2 x *], [2 x *], [1 x 1] -> [1]

>Validating network. 77 nodes to process in pass 2.

The same from python:

>Validating --> __v2libuid__Parameter2010__v2libname__b = LearnableParameter() :  -> [2]
>Validating --> __v2libuid__Plus2023__v2libname__Plus2014 = Plus >(__v2libuid__Times2021__v2libname__Times2012, __v2libuid__Parameter201
>0__v2libname__b) : [2 x *2], [2] -> [2 x *2]

>Validating network. 73 nodes to process in pass 2.

So in python the roots are missing, not that big of a deal, but still weird.

The input and output from GetNodeDimensions seems to be the same, besides the key: 
>Brainscript:    input dimensions: Key: features, value = count: 3072
>Brainscript:    output dimensions = Key: "z", Value = 2 

>python:         input dimensions: Key = "__v2libuid__Input2__v2libname__", Value = 3072
>python:         output dimensions = Key: "__v2libuid__Plus2023__v2libname__Plus2014", value = 2

This is just a test with resnet used on a binary problem.

If I let it run it gets to the evaluate function and at the evalwrapper call:

``
try 
{
        m_eval->Evaluate(stdInputs, stdOutputs)
}
``

Which first throws the error:
![evalerror](https://cloud.githubusercontent.com/assets/3653067/20562161/85510fc6-b182-11e6-9580-c5a2466469e0.png)

If i press continue it throws the exception:
> Error: The output matrix being saved has more data than the numRecords (0) requested to be saved

Again, it works completely fine using a brainscript model and I can't really see any differences in any variables until it throws that error,  so what could be the cause of this?

This was the case with beta 3 and I have update to beta 4 and rebuild the models and the issue is still the same.
Thanks for your help.

**Comment by zhouwangzw** [comment]
@Troelsj Thanks for your experiments. From the ouput of network validation for the Python model, the `__v2libuid__Plus2023__v2libname__Plus2014` should be the output node name. Could you please also show me the code clip around CreateNetwork() and around Evaluate() so that I can have a further investigation? As mentioned you need to explicitly specify the list of output nodes both in CreateNetwork() and Evaluate().
The exception indicates that no output data (numRecords is 0) is generated. What are the Inputs when calling model.Evaluate()?
Sorry for the trouble. We are currently working on C# Eval V2 API which will provide better experience when evaluating models trained by Python. 

Thanks,
Zhou

**Comment by Troelsj** [comment]
@zhouwangzw Yes that is node i used to define the output, so the code around CreateNetwork() looks like:

``
var desiredOutputLayers = new List<string>() { "__v2libuid__Plus2023__v2libname__Plus2014" }; 
model.CreateNetwork(string.Format("traceLevel=1\n modelPath=\"{0}\"", modelFile), deviceId: 0, outputNodeNames: desiredOutputLayers);
``
If i for example change desiredOutputLayers to "z" instead and use a brainscript model it works fine.
 
The code around Evaluate(), resizedCHW is just the loaded bitmap image resized and changed to CHW.
``
var inDims = model.GetNodeDimensions(NodeGroup.Input);
var inputs = new Dictionary<string, List<float>>() { { inDims.First().Key, resizedCHW } };
var outDims = model.GetNodeDimensions(NodeGroup.Output);
outputLayerName = outDims.First().Key;
outputs = model.Evaluate(inputs, outputLayerName);
``
Copy/paste of Inputs in python:
>[0] = {[__v2libuid__Input2__v2libname__, System.Collections.Generic.List`1[System.Single]]}

brainscript:
>[0] = {[features, System.Collections.Generic.List`1[System.Single]]}

So it seems to be the same.

In beta4 you have also implemented a EvaluateRGBImage() function, which I have also tried with the same result (I guess it just automatically convert the image to CHW?), that looks like:

``
var outDims = model.GetNodeDimensions(NodeGroup.Output);
outputLayerName = outDims.First().Key;
outputs = model.EvaluateRgbImage(resized, outputLayerName)
``
Is there any other code snippet you would need?

Looking forward to V2, do you have any estimate of when it will be available?

Thanks

**Comment by zhouwangzw** [comment]
@Troelsj You are currently using the C+ Eval IEvaluateModelManaged interface. We did some investigation and found that the Python model is not supported by this interface. Instead, you can use ModelEvaluationExtendedF() (for float, or ModelEvaluationExtendedD() for double) to evaluate the Python model. I have used this interface to evaluate the model trained by TrainResNet_CIFAR10.py with success. 

The examples for using ModelEvaluationExtended can be found [here](https://github.com/Microsoft/CNTK/blob/master/Tests/UnitTests/ManagedEvalTests/EvalManagedTests.cs). The interface itself is defined [here](https://github.com/Microsoft/CNTK/blob/master/Source/Extensibility/EvalWrapper/EvalExtendedWrapper.cpp). Please specify the output nodes when calling CreateNetwork() as follows:
`var networkConfiguration = "outputNodeNames=\"{1}:{2}\"\n deviceId=-1\n modelPath=\"" + modelFilePath + "\"";
model.CreateNetwork(networkConfiguration);`
Multiple output ndoes need to be separated by ":", as described [here](https://github.com/Microsoft/CNTK/wiki/CNTK-Evaluate-Hidden-Layers).

Thanks

**Comment by Troelsj** [comment]
@zhouwangzw  Okay, that seems to work, is there any plans to make it more streamlined further on?

So I load the model as:

>var model = new ModelEvaluationExtendedF();
>var networkConfiguration = "outputNodeNames=__v2libuid__Plus2023__v2libname__Plus2014\n traceLevel=1\n deviceId=0\n modelPath=" + modelFile + "";
>
>model.CreateNetwork(networkConfiguration);


And evaluate it as:


>VariableSchema inputSchema = model.GetInputSchema();
>VariableSchema outputSchema = model.GetOutputSchema();
>
>model.StartForwardEvaluation(outputSchema.Select(s => s.Name).ToList<string>());
>
>var inputBuffer = inputSchema.CreateBuffers<float>();
>var outputBuffer = outputSchema.CreateBuffers<float>();
>
>int inDims = inputSchema.CreateBuffers<float>()[0].Size;
>for (int i = 0; i < inDims ; i++ )
>{
>        inputBuffer[0].Buffer[i] = resizedCHW[i];
>}
>
>model.ForwardPass(inputBuffer, outputBuffer);
>var decision = outputBuffer[0].Buffer;


I guess that is the correct approach then? Though there probably is a more efficient way of making the image the inputBuffer. 

But otherwise it seems to work, thank you very much for your help :)

**Comment by zhouwangzw** [comment]
Glad to know that it works for you too. Yes, the C# Eval V2 API will provide more convenient way to prepare input data.
Thanks.

**Comment by zhouwangzw** [comment]
I am closing the issue. Please feel free to reopen it if needed.

**Comment by zhouwangzw** [comment]
We have just released the beta version of CNTK Library .NET/C# Eval API. https://github.com/Microsoft/CNTK/releases/tag/v2.0.beta8.0.
